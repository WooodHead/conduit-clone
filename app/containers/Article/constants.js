/*
 *
 * Article constants
 *
 */

export const DEFAULT_ACTION = 'app/Article/DEFAULT_ACTION';
export const FETCH_ARTICLE = 'app/Article/FETCH_ARTICLE';
export const ARTICLE_LOADED = 'app/Article/ARTICLE_LOADED';
export const LOAD_ARTICLE_ERROR = 'app/Article/LOAD_ARTICLE_ERROR';
