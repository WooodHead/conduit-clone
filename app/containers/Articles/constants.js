/*
 *
 * Articles constants
 *
 */

export const FETCH_ARTICLES = 'app/Articles/FETCH_ARTICLES';
export const ARTICLES_LOADED = 'app/Articles/ARTICLES_LOADED';
export const LOAD_ARTICLES_ERROR = 'app/Articles/LOAD_ARTICLES_ERROR';
