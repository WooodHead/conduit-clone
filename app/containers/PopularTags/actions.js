/*
 *
 * PopularTags actions
 *
 */

import {
  DEFAULT_ACTION,
  GET_TAGS,
  TAGS_LOADED,
  TAGS_LOADING_ERROR,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function getTags() {
  return {
    type: GET_TAGS,
  };
}

export function tagsLoaded(payload) {
  const { tags } = payload;
  return {
    type: TAGS_LOADED,
    tags,
  };
}

export function tagsLoadingError() {
  return {
    type: TAGS_LOADING_ERROR,
  };
}
