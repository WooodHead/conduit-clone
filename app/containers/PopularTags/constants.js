/*
 *
 * PopularTags constants
 *
 */

export const DEFAULT_ACTION = 'app/PopularTags/DEFAULT_ACTION';
export const GET_TAGS = 'app/PopularTags/GET_TAGS';
export const TAGS_LOADED = 'app/PopularTags/TAGS_LOADED';
export const TAGS_LOADING_ERROR = 'app/PopularTags/TAGS_LOADING_ERROR';
