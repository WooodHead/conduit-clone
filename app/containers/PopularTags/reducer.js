/*
 *
 * PopularTags reducer
 *
 */

import { fromJS } from 'immutable';
import {
  DEFAULT_ACTION,
  TAGS_LOADED,
  TAGS_LOADING_ERROR,
} from './constants';

const initialState = fromJS({
  fetching: true,
  error: false,
  tags: [],
});

function popularTagsReducer(state = initialState, action) {
  switch (action.type) {
    case DEFAULT_ACTION:
      return state;
    case TAGS_LOADED:
      return state
        .set('fetching', false)
        .set('tags', action.tags);
    case TAGS_LOADING_ERROR:
      return state
        .set('fetching', false)
        .set('error', true);
    default:
      return state;
  }
}

export default popularTagsReducer;
