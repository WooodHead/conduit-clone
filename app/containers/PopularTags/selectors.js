import { createSelector } from 'reselect';

/**
 * Direct selector to the popularTags state domain
 */
const selectPopularTagsDomain = (state) => state.get('popularTags');

/**
 * Other specific selectors
 */


/**
 * Default selector used by PopularTags
 */

const makeSelectFetching = () => createSelector(
  selectPopularTagsDomain,
  (substate) => substate.get('fetching'),
);

const makeSelectError = () => createSelector(
  selectPopularTagsDomain,
  (substate) => substate.get('error'),
);

const makeSelectTags = () => createSelector(
  selectPopularTagsDomain,
  (substate) => substate.get('tags'),
);


export {
  selectPopularTagsDomain,
  makeSelectFetching,
  makeSelectError,
  makeSelectTags,
};
