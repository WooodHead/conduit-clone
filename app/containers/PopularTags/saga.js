import { takeLatest, call, put } from 'redux-saga/effects';
import { GET_TAGS } from 'containers/PopularTags/constants';
import {
  tagsLoaded,
  tagsLoadingError,
} from 'containers/PopularTags/actions';

import request from 'utils/request';

export function* getTags() {
  const apiEndpoint = 'https://conduit.productionready.io/api/tags';
  try {
    const tags = yield call(request, apiEndpoint);
    yield put(tagsLoaded(tags));
  } catch (err) {
    yield put(tagsLoadingError());
  }
}

export default function* tagsData() {
  yield takeLatest(GET_TAGS, getTags);
}
