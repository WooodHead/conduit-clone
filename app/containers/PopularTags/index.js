/**
 *
 * PopularTags
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { push } from 'react-router-redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import { getUrlParams } from 'utils/url';
import { Alert, Button } from 'reactstrap';
import Loader from 'components/Loader';

import { makeSelectFetching, makeSelectTags, makeSelectError } from './selectors';
import reducer from './reducer';
import saga from './saga';
import { getTags } from './actions';

export class PopularTags extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  componentDidMount() {
    this.props.getTags();
  }
  renderButtonTag(tag) {
    const activeTag = getUrlParams(this.props.location.search, 'tag');
    return (
      <Button
        active={tag === activeTag}
        key={`${tag}`}
        outline
        color="primary"
        size="sm"
        onClick={() => {
          this.props.onTagSelect(tag);
        }}
      >
        {tag}
      </Button>
    );
  }
  render() {
    let content = <Loader />;
    if (!this.props.fetching) {
      if (!this.props.error) {
        if (this.props.tags.length !== 0) {
          content = this.props.tags.map((tag) => this.renderButtonTag(tag));
        } else {
          content = (
            <Alert color="info">
              No Tags
            </Alert>
          );
        }
      } else {
        content = (
          <Alert color="danger">
            Something went wrong.
          </Alert>
        );
      }
    }
    return (
      <div>
        <h3>Popular Tags</h3>
        {content}
      </div>
    );
  }
}

PopularTags.defaultProps = {
  fetching: true,
  error: false,
  tags: [],
};

PopularTags.propTypes = {
  fetching: PropTypes.bool.isRequired,
  error: PropTypes.bool.isRequired,
  tags: PropTypes.array.isRequired,
  getTags: PropTypes.func.isRequired,
  onTagSelect: PropTypes.func,
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
};

const mapStateToProps = createStructuredSelector({
  fetching: makeSelectFetching(),
  error: makeSelectError(),
  tags: makeSelectTags(),
});

export function mapDispatchToProps(dispatch) {
  return {
    getTags: () => dispatch(getTags()),
    onTagSelect: (tag) => dispatch(push(`/?tag=${tag}&page=1`)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'popularTags', reducer });
const withSaga = injectSaga({ key: 'popularTags', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(PopularTags);
