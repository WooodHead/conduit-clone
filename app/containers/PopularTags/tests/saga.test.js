/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { GET_TAGS } from 'containers/PopularTags/constants';

import {
  tagsLoadingError,
} from 'containers/PopularTags/actions';
import tagsData, { getTags } from '../saga';

describe('tagsDataSaga Saga', () => {
  let getTagsGenerator;
  beforeEach(() => {
    getTagsGenerator = getTags();

    const selectDescriptor = getTagsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getTagsGenerator.next({ tags: ['foo'] }).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch tagsLoadingError action if api request failed', () => {
    const putDescriptor = getTagsGenerator.throw().value;
    expect(putDescriptor).toEqual(put(tagsLoadingError()));
  });
});

describe('tagsDataSaga Saga', () => {
  const tagsDataSaga = tagsData();

  it('should start task to watch for GET_TAGS action', () => {
    const takeLatestDescriptor = tagsDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(GET_TAGS, getTags));
  });
});
