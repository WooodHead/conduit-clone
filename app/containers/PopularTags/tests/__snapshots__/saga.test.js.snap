// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tagsDataSaga Saga should dispatch tagsLoadingError action if api request failed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://conduit.productionready.io/api/tags",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`tagsDataSaga Saga should dispatch tagsLoadingError action if api request failed 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "tags": Array [
        "foo",
      ],
      "type": "app/PopularTags/TAGS_LOADED",
    },
    "channel": null,
  },
}
`;
